contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,850 @@
+     `.Status.TemplateRevision`.
+    1. All Pods with ordinals in the sequence 
+     `[Spec.RevisionParition,.Spec.Replicas)` have been created with the 
+     PodTemplate in the StatefulSet's revision history that is labeled with a 
+     `StatefulSetTemplateRevisionLabel` equal to 
+     `.Status.TargetTemplateRevision`.
+
+### Revised Controller Algorithm
+The StatefulSet controller will use the following algorithm to continue to 
+make progress toward the user declared [target state](#target-state) while 
+respecting the controller's 
+[identity](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#pod-identity), 
+[deployment, and scaling](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#deployment-and-scaling-guarantee) 
+guarantees.
+
+1. The controller will 
+[reconstruct the revision history](#history-reconstruction) of the StatefulSet.
+1. The controller will process any [template updates](#template-updates) to 
+ensure that the StatefulSet's revision history is consistent with the user 
+declared desired state.
+1. The controller will select all Pods in the StatefulSet, filter any Pods not 
+owned by the StatefulSet, and sort the remaining Pods in ordinal order.
+1. If any Pods with ordinals in the sequence `[0,.Spec.Replicas)` have not been 
+created, for the Pod corresponding to the lowest such ordinal, the controller 
+will [select the PodTemplate](#podtemplate-selection) from the StatefulSet's 
+revision history corresponding to the Pod's ordinal and create the Pod.
+1. If all Pods in the sequence `[0,.Spec.Replicas)` have been created, but if any 
+have a Status other than Running or do not have a Ready Condition, the 
+StatefulSet controller will wait for these Pods to either become Running and 
+Ready, or to be completely deleted. 
+1. If all Pods in the sequence `[0,.Spec.Replicas)` have a Status of Running and 
+a Condition indicating Ready, and if `.Spec.Replicas` is less than 
+`.Status.Replicas`, the controller will delete the Pod corresponding to the 
+largest ordinal. This implies that scaling takes precedence over Pod updates.
+1. If all Pods in the range `[0,.Spec.Replicas)` have a Status of Running and 
+a Ready Condition, if `.Spec.Replicas` is equal to `.Status.Replicas`, and if 
+there are Pods that do not match the 
+[declared desired PodTemplate](#podtemplate-selection), the Pod corresponding to 
+the largest ordinal will be deleted.
+1. If the StatefulSet controller has achieved the 
+[declared target state](#target-state), and if that state has a 
+`.Spec.ParitionOrdinal` equal to `0`, the StatefulSet controller will 
+[complete any in progress updates](#update-completion).
+1. The controller will [report its status](#status-reporting).
+1. The controller will perform any necessary
+[maintenance of its revision history](#history-maintenance).
+
+### StatefulSet Revision History
+The StatefulSet controller will use labeled, versioned PodTemplates to keep a 
```

Comment:
So for pod template here's what we need to handle:

1. Collisions across controllers by naming
2. Idempotency of product versions (the same pod template in a controller object should map to exactly one PodTemplate across multiple upgrades)
3. Human readability - a human should be able to find a PodTemplate
4. Collisions across controllers by pod template spec - if two controllers deploy the exact same pod template (daemon set vs stateful set) the PodTemplate has two owners, not just one, so garbage collection won't work
5. Future versions of controllers may involve renames of the controller resource name, so we would still have to be able to locate old PodTemplate revisions even if the resource name changes
6. Immutability of PodTemplate - a template used in this fashion needs to prevent updates, or at least strongly discourage them (if they are mutable, we can't compare the pods against the pod template version).